Feat/#19 - 간단한 JWT 인증/인가 구현

Sunja-An opened the pull request:
> [!WARNING]
> **이슈와 프로젝트(Project) 연결 시 반드시 주의하세요.**
> 이는 업무 효율도를 위해서 사용하는 것이지 때문에 협조 부탁드립니다. 추가로, 완료 된 PR 이라면 리뷰어를 직접적으로 연결해주세요 :)

## 📌 연결된 이슈
- resolve: #19 

## 📖 주요 변경 사항
- Domain 에 들어가있던, AccessToken, RefreshToken 은 Infra 레이어에서 Jwt 속성에 대한 값을 담을 수 있는 Persistence Record 객체로 설정하였습니다.
- 아직 Test 코드 작성 안했어요!!!

## 🔖 작업 내용 요약
<img width="1920" height="1080" alt="Board" src="https://github.com/user-attachments/assets/fb9388b4-7987-4a8f-b5be-007fb0a43c83" />

**[위의 그림을 참고해서 봐주세요！]**

- TokenProperties 를 통해서 Token 에 대한 속성들을 application.yaml 에서 가져옵니다. ( 값은 env 로 관리 )
- 부여하는 과정은 TokenProvider 생성자에 부여해놨습니다.

- TokenProvider 를 통해 Token 에 대한 속성들을 담고있는 Record 객체들에 대해서 접근합니다.

- 이렇게 만듦으로써, JwtProvider 에서 TokenType 에 따라서 속성을 가져와, Token 을 발행해줄 수 있습니다.
- 또한 토큰 발행과 토큰 속성 부여 책임을 분리함으로써, Error 가 일어났을 때, 추적하기 쉽도록 파일 구조를 선언하였습니다.

<img width="1075" height="821" alt="image" src="https://github.com/user-attachments/assets/9010bcde-24a2-4cb8-8935-21c0ae23f680" />

---
### 2차 PR ( 11.26(水) )

<img width="1920" height="1080" alt="2025 11 26" src="https://github.com/user-attachments/assets/943b8bcb-6948-4dc6-ab16-a025aa46dd78" />

- Jwt Provider 에서의 Token 분해 및 정보 가져오기를 삭제하였습니다.
- 정보 가져오는 책임은 AOP 를 이용한 Resolver 나 aspect 에서 담당하고 있습니다.
- 또한 Token 분해 및 검증에 대한 책임은 Filter 에서 처리할 수 있도록, Utils 을 설정하고, ErrorCode 를 반환할 수 있도록 설정하였습니다.

- @PreAuthorize 어노테이션의 경우, String 으로 Role 에 대한 내용을 받아오고, 해당 Role 에 대해서는 UserRole.java 파일 ( ENUM 파일 ) 에서 검증을 할 수 있도록 코드를 작성하였습니다.

- 추가적으로 @AuthUser 어노테이션과 분리를 하였습니다만, 이유는 Token 의 Subject 정보 ( User ID ) 를 원하는 Method 와 검증만을 원하는 Method 가 따로 있을 것이라고 생각하여, 분리를 하게 되었습니다.

Sunja-An committed:
Feat: .env file added to gitignore (#19)

Sunja-An committed:
Feat: ErrorCode updated (#19)

Sunja-An committed:
Feat: UserRole added to User Entity (#19)

Sunja-An committed:
Feat: Method that can find by email value is added (#19)

Sunja-An committed:
Feat: application.yaml added for jwt properties (#19)

Sunja-An committed:
Feat: JwtProperties added and root application can scan it (#19)

Sunja-An committed:
Feat: Util method added for Date Object (#19)

Sunja-An committed:
Feat: Token Object added for store token information (#19)

Sunja-An committed:
Feat: DTO for sending Token information (#19)

Sunja-An committed:
Feat: TokenProvider that can get secret key or ttl is added (#19)

Sunja-An committed:
Feat: JwtService interface is added for connecting to Infra layer component (#19)

Sunja-An committed:
Feat: JwtProvider added for publishing token (#19)

Sunja-An committed:
Feat: DTO is added for AuthController (#19)

Sunja-An committed:
Feat: Controller and Service related with auth are added (#19)

BaeJunH0 commented on src/main/java/com/example/bak/BakApplication.java:
```diff
@@ -2,8 +2,10 @@
+import org.springframework.boot.context.properties.ConfigurationPropertiesScan;
 
 @SpringBootApplication
+@ConfigurationPropertiesScan
```
SpringApplication 클래스는 최대한 퓨어하게 하기로 했습니다! (컨벤션에 추가해둘게요!)
따로 Configuration 클래스 만들어서 분리해주심 감사하겠습니다.

Sunja-An committed:
Feat: UserRepositoryStub findByEmail Method Test added (#19)

Sunja-An committed:
Feat: UserRole convert method modified (#19)

Sunja-An committed:
Feat: yaml file convention modified (#19)

Sunja-An committed:
Feat: Indent added easy to read (#19)

Sunja-An committed:
Feat: Wrong convention modified (#19)

BaeJunH0 commented on src/main/java/com/example/bak/auth/application/JwtService.java:
```diff
@@ -0,0 +1,10 @@
+import com.example.bak.auth.infra.jwt.persistence.TokenType;
+import com.example.bak.user.domain.UserRole;
+
+public interface JwtService {
```
인프라 구현체와 분리를 위한 부분인듯한데, 뭔가 Service의 구현체인 Provider는 와닿지 않는 느낌이 있습니다 (개인적 의견)
차라리 Service - ServiceImpl이나, Provider-ProviderImpl, 아니면 아얘 정석적으로 Port - Adaptor 느낌의 작명은 어떨까요?

BaeJunH0 commented on src/main/java/com/example/bak/auth/infra/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,52 @@
+@RequiredArgsConstructor
+public class JwtProvider implements JwtService {
+    private final TokenProvider tokenProvider;
+    private final static String tokenRoleKey = "role";
```
final static 타입 변수는 대문자 스네이크 케이스로 적는게 컨벤션에 맞을 것 같습니다!
https://github.com/JunHoPark93/google-java-styleguide <- 5.2.4 문단 참고해주세요!

BaeJunH0 commented on src/main/java/com/example/bak/auth/infra/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,28 @@
+package com.example.bak.auth.infra.jwt;
+
+import com.example.bak.auth.infra.jwt.persistence.*;
```
와일드카드 사용하면 안됩니당
https://github.com/JunHoPark93/google-java-styleguide <- 3.3.1 문단 참고해주세요!

Sunja-An replied:
넵! 감사합니다!
참고해서 고쳐놓겠습니다 😭

BaeJunH0 commented on src/main/java/com/example/bak/auth/infra/jwt/persistence/AccessToken.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.bak.auth.infra.jwt.persistence;
+
+public record AccessToken(String secretKey, Long timeToLive) implements Token {
```
아까 설명하신대로 뭔가 프로퍼티에 대한 클래스라는 것을 명시하지 않으면 크게 혼동이 올 것 같습니다.
네이밍을 바꾸는게 어떠할지...

Sunja-An replied:
추후 AccessTokenProperties 와 같이 명시해놓고자 합니다!!

BaeJunH0 commented on src/main/java/com/example/bak/global/common/utils/DateUtils.java:
```diff
@@ -0,0 +1,29 @@
+
+@UtilityClass
+public class DateUtils {
+    public final static ZoneId zoneId = ZoneId.of("Asia/Seoul");
+
+    public static Date getNow() {
+        Instant nowUtc = Instant.now();
+        ZonedDateTime now = nowUtc.atZone(zoneId);
+        return Date.from(now.toInstant());
+    }
+
+    public static Date getExpirationDate(Long ttl){
+        if(ttl == null){
+            return null;
+        }
+        Instant nowUtc = Instant.now().plusMillis(ttl);
+        ZonedDateTime now = nowUtc.atZone(zoneId);
+        return Date.from(now.toInstant());
+    }
+
+}
```
LGTM 👍

BaeJunH0 commented on src/main/java/com/example/bak/global/exception/ErrorCode.java:
```diff
@@ -14,8 +14,12 @@
+
     // User
     USER_NOT_FOUND("US001", HttpStatus.NOT_FOUND, "사용자 리소스를 찾을 수 없습니다."),
+    CANNOT_CONVERT_ROLE("US002", HttpStatus.INTERNAL_SERVER_ERROR, "올바르지 않은 Parameter 입력입니다."),
```
500번대 에러는 서버측 에러로, 저희가 예상하지 못한 부분에만 공통으로 사용해야 한다고 알고 있는데... 
매개변수 입력은 서버의 책임일까요?

BaeJunH0 commented on src/main/java/com/example/bak/user/domain/UserRole.java:
```diff
@@ -0,0 +1,20 @@
+        }else if(role.equalsIgnoreCase("ADMIN")) {
+            return ADMIN;
+        }
+        throw new IllegalArgumentException(CANNOT_CONVERT_ROLE.getMessage());
```
BussinesException이 아닌 IllegalArgumentException을 사용한 이유가 있을까요?

Sunja-An replied:
해당 사안에 대해서 깊게 생각하지 않았습니다..
수정 해놓도록 하겠습니다!!

BaeJunH0 commented on src/main/java/com/example/bak/user/domain/User.java:
```diff
@@ -42,6 +41,12 @@ private User(Long id, String email, String password) {
         this.password = password;
     }
 
+    public void matchPassword(String newPassword) {
+        if (!newPassword.equals(this.password)) {
+            throw new BusinessException(ErrorCode.INCORRECT_PASSWORD);
+        }
+    }
```
도메인 로직에서 비즈니스 예외 사항을 잡는 느낌이 듭니다.
도메인 로직에서는 같은지, 아닌지 판단을 하고, 비즈니스 로직으로 올려서 exception 여부를 결정하는 것은 어떨까요?
요거는 코딩 스타일 일 수도 있어서 의견만 주셔도 됩니당.

Sunja-An replied:
이것도 추후 반영을 해놓겠습니다!
늦게 확인해서 죄송합니다!

BaeJunH0 commented on src/main/resources/application.yaml:
```diff
@@ -13,4 +15,13 @@ spring:
+  auth:
+    access:
+      secret-key: ${JWT_ACCESS_KEY}
+      timeToLive: ${JWT_ACCESS_EXPIRATION}
```
time-to-live로 변경된 것 아니었나용

Sunja-An replied:
해당 부분은 변경하였습니다!

BaeJunH0 reviewed (changes requested):
컨벤션 조금씩 나간 부분이랑, 질문 사항들 있습니다! 천천히 답변 주세요~😊👌

Sunja-An committed:
Feat: 인증 관련 ErrorCode 추가 (#19)
- 토큰이 없는 경우
- 토큰이 부적절한 경우
- 원하는 Method 를 실행할 권한이 없는 경우
- Token 을 분해했지만, 정보가 없는 경우

Sunja-An committed:
Feat: Jwt 로직 중 사용 여지가 없는 Method 삭제 (#19)

Sunja-An committed:
Feat: 인가를 위한 UserRole ENUM 추가 (#19)

Sunja-An committed:
Feat: 인증/인가 관련한 Custom Exception 추가 (#19)

Sunja-An committed:
Feat: Jwt Token Validation Util Method 추가 (#19)

Sunja-An committed:
Feat: AuthUser Annotation, Resolver 추가 및 Resolver WebMvcConfig 에 등록 (#19)

Sunja-An committed:
Feat: Jwt Token 검증 필터 추가 (#19)

Sunja-An committed:
Feat: 단일책임원칙 (SRP)에 따른 Jwt Validator 추가 (#19)

Sunja-An committed:
Feat: 인증 관련 정보를 담는 DTO 추가 (#19)

Sunja-An committed:
Feat: 단순 인증만을 부여할 수 있는 Annotation 과 AOP 적용 (#19)

Sunja-An committed:
Feat: Token 과 관련된 DTO 역할에 맞는 네이밍 변경 (#19)

Sunja-An committed:
Feat: import 부분 와일드카드 적용 부분 삭제 (#19)

Sunja-An committed:
Feat: Port&Adaptor 선정으로 인한, 네이밍의 모호함 수정 (#19)

Sunja-An committed:
Feat: 사용자의 입력에 따른 변환 불가 사항에 대해서 BAD REQUEST 반환하도록 수정 (#19)

Sunja-An committed:
Feat: Illegal Exception 에서 BusinessException 으로 변환(#19)

Sunja-An committed:
Feat: Gradle 버젼 추가 (#19)

Sunja-An committed:
Feat: JwtProperties Setter 주입으로 인한 Record -> Class 변경. 변경 후, Getter 로 가져오는 방식 변경 (#19)

Sunja-An committed:
Feat: PropertiesScan 어노테이션 삭제 (#19)

Sunja-An committed:
Feat: application-local, prod 에 Jwt 설정 추가 (#19)

Sunja-An committed:
Fix: application.yaml 삭제 및 CommandLineRunner 삭제 (#19)

Sunja-An committed:
chore: format 형식 수정 (#19)

polyglot-k commented on src/main/java/com/example/bak/auth/application/AuthService.java:
```diff
@@ -0,0 +1,30 @@
+package com.example.bak.auth.application;
+
+import com.example.bak.auth.application.dto.TokenInfo;
+import com.example.bak.auth.infra.jwt.persistence.TokenType;
```
application 에서 infra layer 를 직접 참조 하고 있네요 :) ☺️
이 부분에 대해서 수정해주세요

polyglot-k commented on src/main/java/com/example/bak/auth/application/AuthService.java:
```diff
@@ -0,0 +1,30 @@
+        user.matchPassword(password);
+
+        String accessToken = jwtPort.publishToken(TokenType.ACCESS, user.getId(), user.getRole());
+        String refreshToken = jwtPort.publishToken(TokenType.REFRESH, user.getId(), user.getRole());
```
이런 TokenType 은 오히려 Domain 적인 요소 같다는 생각이 들어요.

jwt 라는 것을 추상화 한 것이 TokenType인 Refresh, Access 라고 생각이 드네요!

infra 에서 application 이나 domain 을 참조 하는 것은 허용 되서 이부분에 대해서 수정 부탁드립니다.

polyglot-k commented on src/main/java/com/example/bak/auth/application/JwtTokenPort.java:
```diff
@@ -0,0 +1,8 @@
+import com.example.bak.user.domain.UserRole;
+
+public interface JwtPort {
+    String publishToken(TokenType tokenType, Long id, UserRole role);
```
이거 가독성이 조금 떨어져서, 원래는 DTO 를 넣어서, publish 에 대한 커맨드를 보내지만, 저희는 DTO 최소형태를 띄는 것을 지향하기 때문에 여기는 주석 javadoc 으로 인자가 뭘 가르키는지 설명 부탁드려요

polyglot-k commented on src/main/java/com/example/bak/auth/infra/jwt/JwtAdaptor.java:
```diff
@@ -0,0 +1,27 @@
+
+@Component
+@RequiredArgsConstructor
+public class JwtAdaptor implements JwtPort {
```
JwtAdaptor 랑 JwtPort 라는 느낌보다는 TokenGeneratorPort, TokenGeneratorAdaptor 로 하고, Adaptor에서 Jwt 를 사용하는 형태로 네이밍에 대한 추상화를 조금 더 올리면 좋을 것 같다는 생각이 듭니다 😎

polyglot-k commented on src/main/java/com/example/bak/auth/infra/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,33 @@
+
+    public SecretKey getSecretKey(TokenType tokenType) {
+        return Keys.hmacShaKeyFor(Decoders.BASE64.decode(tokens.get(tokenType).secretKey()));
+    }
```
```suggestion
    public SecretKey getSecretKey(TokenType tokenType) {
        타입 시크릿 = Decoders.BASE64.decode(tokens.get(tokenType).secretKey());
        return Keys.hmacShaKeyFor(분리한 내용);
    }
```

이런 형태로 수정을 하는게 추후에 알아보기 더 편할 것 같은데 @Sunja-An 님은 어떻게 생각하세요?

Sunja-An replied:
해당 의견에 대해서는 저도 동의를 하는 부분입니다!
1줄 코드로 적는게 멋있다는 강박관념이 아직 남아 있는 것 같네요 ㅎㅎ..😔

BaeJunH0 commented on src/main/java/com/example/bak/auth/infra/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Map;
+
+@Component
+public class TokenProvider {
```
Infra에서 사용하는 부분이고, 구체적인 기술 종속성을 띄는 구현체니까 차라리 JwtTokenProvider 이런 식으로 하는게 직관적일 것 같습니다!

아니면 아마 그럴 일은 없겠지만 Token 발행 종속성을 바꾼다면 TokenProvider를 interface로 두게되면 좀 더 확장성 있는 설계가 될 것 같기도 합니다.

polyglot-k commented on src/main/java/com/example/bak/auth/infra/jwt/utils/JwtUtils.java:
```diff
@@ -0,0 +1,24 @@
+    public static String resolveToken(String bearerToken) {
+        if(StringUtils.hasText(bearerToken) && bearerToken.startsWith(PREFIX_BEARER_TOKEN)) {
+            return bearerToken.substring(PREFIX_BEARER_TOKEN.length());
+        }
```
[개인적인 궁금증] 그냥 생각을 했는데 StringUtils.hashText 로 판단하는거보다 bearerToken이 null 인거만 체크를 하면 나머지에서는 충분히 괜찮지 않나요?? hasText를 꼭 필요로 하는 이유가 궁금합니다.

Sunja-An replied:
이거는 전적으로 동의합니다.
수정 적용해놓도록 하겠습니다!

polyglot-k commented on src/main/java/com/example/bak/auth/infra/jwt/utils/JwtUtils.java:
```diff
@@ -0,0 +1,24 @@
+        if(StringUtils.hasText(bearerToken) && bearerToken.startsWith(PREFIX_BEARER_TOKEN)) {
+            return bearerToken.substring(PREFIX_BEARER_TOKEN.length());
+        }
+        return null;
```
null 을 반환하는 형태를 Optional 없이 하신 이유가 있으신지 궁금합니다. 비즈니스 레이어에서 null 을 제대로 처리하지 않음으로 인해서 NPE 가 발생할 수 있을 가능성이 있을 것 같은데, 어떤 철학으로 이 부분을 이런식으로 했는지 공유해주시면 좋을 것 같습니다 :)

Sunja-An replied:
Optional 객체를 생각을 하지 못했습니다!
평소 Optional 을 쓸 때에는 객체를 가져올 때, null 이 있을 때, 위험한 DB 에서 가져오는 객체들에 대해서만 사용하는 습관을 가지고 있던 것 같습니다!
이 부분에 대해서는 수정을 해놓도록 하겠습니다!

Sunja-An replied:
더 고민을 해보았을 때, Optional 을 쓰게 된다면 저는 아래와 같이 작성이 됩니다.
```java
Optional<String> token  = JwtUtils.resolveToken(request.getHeader(HttpHeaders.AUTHORIZATION));
if(token.isEmpty()){
  filterChain.doFilter(request, response);
  return;
}
tokenValue = token.get();
```
해당 방식보다는 제가 한 방식이 조금 더 가독성이 있고, 직관적이다 라는 생각이 드는데 어떨지 여쭤보고 싶습니다!

polyglot-k commented on src/main/java/com/example/bak/global/common/utils/DateUtils.java:
```diff
@@ -0,0 +1,29 @@
+    public static Date getExpirationDate(Long ttl){
+        if(ttl == null){
+            return null;
+        }
```
ttl 이 음수인 케이스는 통과해서 동작하는게 맞을까요?? 이 부분은 고려해보셨을까요?

Sunja-An replied:
이 부분에 대해서는 Environment 으로 TTL 을 적용해서, 해당 부분에 있어서는 음수케이스를 Test Case 에 등록하지 않았습니다!
만약 음수가 들어왔을 경우, 토큰이 만료가 바로 되는 Error 로 이어져, 시스템 상 다운이 되는 오류까지는 안 갈 것이라고 생각합니다!
( 근데 지금 생각해보니, 이게 더 심각한 오류인 것 같기도 하구요... 바로 파악이 안되니... )

Sunja-An replied:
더 생각을 해보았을 때, Token 의 TTL 이 음수로 들어오는 상황은 서비스가 돌아가면 안되는 상황이니, Exception 으로 처리해도 괜찮을 것 같다는 생각을 해보았습니다!
따라서 다음 PR 에 적용해놓겠습니다!

BaeJunH0 commented on src/main/java/com/example/bak/global/security/filter/JwtTokenRequestFilter.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        String requestURL = request.getRequestURI();
+        if(!isWhiteList(requestURL)) {
```
이중 부정으로 뭔가 로직이 꼬인 것 같습니다.

WhiteList가 아닌 requestUrl에 대해서 다음 로직이 진행되어야 할 것 같은데, 해당 if문에 whiteList가 아닌 url이 걸리는 상황 같습니다.

Sunja-An replied:
이중 부정으로 적용된 부분을 PatternStringUtils 의 SimpleMatch 는 만약 String[] 내부에 대상 String 이 들어가있으면, return true 를 하는 방향으로 되어있어, 부정문인 ! 를 제외하였습니다!

좋은 지적 감사합니다!

polyglot-k commented on src/main/java/com/example/bak/global/security/filter/JwtTokenRequestFilter.java:
```diff
@@ -0,0 +1,58 @@
+
+@Component
+public class JwtTokenRequestFilter extends OncePerRequestFilter {
+    private static final String[] whiteList = {"/api/v1/users", "/api/v1/auth/login", "/css/*"};
```
whiteList 가 조금도 메소드 단위로, 세분화 될 수도 있지 않을까요? 당장에는 url 형태만을 whiteList 로 관리한거 같은데, 그리고 상수라면 네이밍을 대문자 형태로 WHITE_LIST 가 더 적절할 것 같습니다.

Sunja-An replied:
상수로 되어있을 때, 네이밍을 대문자 형태로 설정하는 것에는 동의합니다!
세분화라는 개념(?)에 맞는지는 잘 모르겠지만, 만약 이 코드를 다시 수정하게 된다면
```java
private static final ArrayList<String> whiteList = new ArrayList<>();
```
위의 코드를 설정한 후, Method 별로 whiteList.add(""); 형식처럼 들어가도 괜찮을 것 같은데, 이 방식은 괜찮을까요?

polyglot-k commented on src/main/java/com/example/bak/global/security/aspect/PreAuthorizeAspect.java:
```diff
@@ -0,0 +1,41 @@
+package com.example.bak.global.security.aspect;
+
+import com.example.bak.global.exception.ErrorCode;
+import com.example.bak.global.exception.VerificationException;
+import com.example.bak.global.security.annotation.PreAuthorize;
+import com.example.bak.global.security.dto.AuthInfo;
+import com.example.bak.auth.infra.jwt.utils.JwtUtils;
+import com.example.bak.user.domain.UserRole;
+import jakarta.servlet.http.HttpServletRequest;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+@Component
+@Aspect
+public class PreAuthorizeAspect {
+    @Around("@annotation(preAuthorize)")
+    public Object checkRoleForAuthorize(ProceedingJoinPoint pjp, PreAuthorize preAuthorize) throws Throwable {
+        ServletRequestAttributes attrs = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
+        if (attrs == null) {
+            throw new VerificationException(ErrorCode.NO_ATTRIBUTES);
+        }
+
+        HttpServletRequest request = attrs.getRequest();
+
+        AuthInfo auth = (AuthInfo) request.getAttribute(JwtUtils.CLAIMS_KEY);
+        if(auth == null){
+            throw new VerificationException(ErrorCode.NO_TOKEN);
+        }
+
+        UserRole standardRole = UserRole.convert(preAuthorize.role());
+
```
이 부분은 그냥 개인적인 생각인데 method 단위로 분리를해서, step by step 이 조금 더 보이게 하면 좋을 것 같습니다.
메소드 한개에 대한 책임이 큰거 같은데, 로직을 추후 확장함에 있어서 조금 힘들 수도 있어보입니다.

논리 흐름 자체는 괜찮은 것 같습니다. 그리고 instanceOf 로 조금 더 안정성있게 java 16+에서 사용되는 문법으로 변환할 수 있어보이는데 여기에 대해서는 어떻게 생각하세요!? 🫢

Sunja-An replied:
instanceof 을 까먹고 있었네요!!
해당 문법이 들어가면 코드가 더 이쁠 것 같습니다!

혹은 AuthInfo 클래스 내부에서 Optional 로 반환해서, null 로 바뀌는 과정들을 생략해서 조금 더 깔끔한 코드를 작성할 수 있겠다 라는 생각도 드네요! ( 이 부분 어떻게 생각하세요 ?? )

책임에 대한 분리는 동의합니다..조금 더 머리 짜서 코드 작성해보겠습니다!

BaeJunH0 commented on src/test/java/com/example/bak/user/domain/UserRepositoryStub.java:
```diff
@@ -22,4 +22,9 @@ protected boolean isSame(Long left, Long right) {
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        return store.stream().filter(u -> Objects.equals(u.getEmail(), email)).findFirst();
```
```suggestion
        return store.stream()
                .filter(u -> Objects.equals(u.getEmail(), email))
                .findFirst();
```

메서드 체이닝 과정에서는, 특히 stream이면 이후 체이닝은 줄을 바꿔주는게 가독성이 좋습니다!

polyglot-k commented on src/main/java/com/example/bak/global/security/filter/JwtTokenRequestFilter.java:
```diff
@@ -0,0 +1,58 @@
+
+    public JwtTokenRequestFilter(JwtValidator jwtValidator) {
+        this.jwtValidator = jwtValidator;
+    }
```
이런 형태로 주입받는 의도가 궁금합니다! 저희 팀 내에서는 `@RequiredArgsConstructor` 를 적극적으로 쓰고 있는데, 의도가 있을까요?

Sunja-An replied:
헉.. 이 부분에 대해서는 JwtValidator 가 의존성 주입으로 들어오는 것을 확실하게 받아오고 싶다(?) 는 생각으로 적용하게 되었는데, 의도는 @RequireArgsConstructor 를 썼을 때, 과연 이게 제대로 들어올까..? 라는 의문점을 가지고 했기 때문에 확실한 방법인 생성자 주입을 선택하였습니다..
팀 컨벤션에 대한 내용이 많은 관계로 이 부분은 계속해서 수정해 나가도록 하겠습니다!

polyglot-k commented on src/main/java/com/example/bak/global/security/filter/JwtTokenRequestFilter.java:
```diff
@@ -0,0 +1,58 @@
+            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, ErrorCode.INVALID_TOKEN.getMessage());
+            return;
+        }
+        filterChain.doFilter(request, response);
```
method 분리를 이부분도 조금 더 진행하셔도 될 것 같습니다. 한 메소드의 책임이 너무 강합니다.

BaeJunH0 commented on src/main/java/com/example/bak/global/exception/ErrorCode.java:
```diff
@@ -14,12 +14,19 @@
+    INCORRECT_PASSWORD("AU001", HttpStatus.UNAUTHORIZED, "비밀번호가 일치하지 않습니다"),
+    INVALID_TOKEN("AU002", HttpStatus.UNAUTHORIZED, "토큰이 유효하지 않습니다"),
+    NO_ATTRIBUTES("AU003", HttpStatus.BAD_REQUEST, "토큰 검증에 필요한 정보가 없습니다."),
+    NO_TOKEN("AU004", HttpStatus.UNAUTHORIZED, "토큰이 없습니다."),
+    NO_AUTHORITY("AU005", HttpStatus.FORBIDDEN, "해당 정보에 접근할 수 있는 권한이 없습니다."),
     // Company
```
```suggestion
    NO_AUTHORITY("AU005", HttpStatus.FORBIDDEN, "해당 정보에 접근할 수 있는 권한이 없습니다."),
    
    // Company
```

공백 한 칸 들어가면 구분되고 참 좋을 것 같아용

polyglot-k commented on src/main/java/com/example/bak/user/domain/UserRole.java:
```diff
@@ -0,0 +1,26 @@
+
+import static com.example.bak.global.exception.ErrorCode.*;
+
+public enum UserRole {
+    NORMAL,
+    ADMIN;
+
+    public static UserRole convert(String role) {
+        if(NORMAL.name().equalsIgnoreCase(role)) {
+            return NORMAL;
+        }else if(ADMIN.name().equalsIgnoreCase(role)) {
+            return ADMIN;
+        }
+        throw new BusinessException(CANNOT_CONVERT_ROLE);
+    }
```
깔끔하고 좋습니다 ☺️☺️

polyglot-k commented on src/main/java/com/example/bak/global/security/resolver/AuthUserArgumentResolver.java:
```diff
@@ -0,0 +1,35 @@
+        AuthInfo authInfo = (AuthInfo) httpServletRequest.getAttribute(JwtUtils.CLAIMS_KEY);
+        if(authInfo == null) {
+            return null;
+        }
```
이거 필요로 하면, 사용하는 형태인데, 없다면 null 을 넘기는게 맞을까요? 에러를 퍼트려야할까요? 이부분은 한번 고려해보면 좋을 것 같아요. controller에 null이 넘어간다는건 굉장히 위험한 것 같아서 여쭤봅니다.

`@PathVariable` 과 같은 어노테이션처럼 requires 같은 옵션이나 nullable 에 대한 옵션을 어노테이션에서 지원하는게 더 안전한 형태를 띌 것 같습니다.

Sunja-An replied:
해당 부분은 null 을 넘기는게 아니라, Response 로 해당 Resolver 에서 에러를 바로 반환하는게 옳은 방식인 것 같습니다!!

또한 @AuthUser 에 requires 옵션을 다는 것도 고려해보겠습니다!
굉장히 재밌을 것 같네요!

정성스러운 리뷰에 정말 감사드립니다!!!!!

polyglot-k reviewed (changes requested):
전반적으로 고민한 흔적이 잘 보이고, 전체적인 흐름도 잘 파악할 수 있어서 좋았습니다.
다만, 고도화가 진행될수록 메서드가 조금 비대해질 가능성이 있어 보였고, 일부는 의도를 바로 이해하기 어려운 부분도 있었습니다.

레이어 간 경계나 의존성 흐름 같은 부분만 조금 더 신경 써주시면 훨씬 더 깔끔해질 것 같습니다.
작성해주신 코드 덕분에 전체적인 구조를 따라가기 편했고, 노력해주신 점 감사드립니다!

부분적인 수정이나 리팩터링이 필요해 보이는 부분은 issue 로 확장해주시면, 그 이후에 다시 리뷰 요청 주시면 Approve 드리겠습니다. 🙌

BaeJunH0 commented on src/main/java/com/example/bak/global/security/resolver/AuthUserArgumentResolver.java:
```diff
@@ -0,0 +1,35 @@
+            return null;
+        }
+
+        return authInfo.id();
```
단순히 null이 반환되는 것 보다는, Exception이 발생해야 하지 않나? 싶은 부분입니다.

왜냐면, @AuthUser로 id나 role에 대한 접근을 필요로 하는 controller 부분에서 null이 반환되는 것을 처리하는 것 보다는, 해당 부분에서 exception으로 처리하는 게 좀 더 격리적으로 잘 처리할 수 있을 거 같은 부분입니다.

Sunja-An replied:
해당 코멘트에 대해서는 적용하겠습니다!

좋은 의견 감사합니다 ☺️

Sunja-An committed:
fix: 이중 부정문 수정 (#19)

Sunja-An committed:
fix: DateUtils 에서 진행하던 ttl 에 대한 예외처리를 Utils 에 책임부여했던 것을 Fix (#19)

Sunja-An committed:
Fix: Domain 구조 변경 (#19)
- JWT 토큰에 대한 객체에 대해 참조를 하기 위해서는 Infra 레이어가 Application 에 묻어있어야 했음.
- 따라서 JwtToken 이라는 interface 와 AccessToken, RefreshToken 이 해당 interface 를 구현하여, 추후 전략패턴을 사용하여 TokenType 에 따른 객체를 가져와 토큰을 발행하기 위함.

Sunja-An committed:
Fix: JwtUtils 내부 로직 변경 (#19)
- Optional 객체 사용
- 전달받은 Param 값 null 여부 확인 로직 수정

Sunja-An committed:
Feat: ErrorCode 수정 (#19)
- 토큰에 대한 환경설정 ErrorCode 추가

Sunja-An committed:
fix: 내부 로직 Method 분리

Sunja-An committed:
fix: 내부 로직 Method 분리 (#19)

Sunja-An committed:
feat: 파일명 변경 및 전략패턴을 적용한 Token Provider 추가(#19)

Sunja-An committed:
feat: Method분리 및 인터페이스 파일명 변경 (#19)

Sunja-An committed:
feat: JwtTokenPort Javadoc 추가 (#19)

BaeJunH0 commented on src/main/java/com/example/bak/auth/application/dto/TokenInfo.java:
```diff
@@ -0,0 +1,10 @@
+package com.example.bak.auth.application.dto;
+
+public record TokenInfo(
```
Application -> Presentation으로 정보를 전달하는 DTO의 경우에, Result 네이밍을 사용하기로 했습니다!

BaeJunH0 commented on src/main/java/com/example/bak/auth/domain/JwtToken.java:
```diff
@@ -0,0 +1,10 @@
+package com.example.bak.auth.domain;
+
+public sealed interface JwtToken permits AccessToken, RefreshToken {
+
+    String getSecretKey();
+
+    Long getTimeToLive();
+
+    TokenType getTokenType();
+}
```
LGTM ❤

BaeJunH0 commented on src/main/java/com/example/bak/auth/infra/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,32 @@
+@Component
+public class JwtTokenProvider {
+
+    private final Map<TokenType, JwtToken> tokens = new EnumMap<>(TokenType.class);
+
+    public JwtTokenProvider(List<JwtToken> tokens) {
+        for (JwtToken token : tokens) {
+            this.tokens.put(token.getTokenType(), token);
+        }
+    }
```
LGTM 👍

BaeJunH0 reviewed (approved):
리뷰 반영 잘 해주셨습니다! 

특히 sealed 사용해서 List<JwtToken> 방식으로 Bean 주입하는 부분 좀 섹시하네요 ㅎㅎ

다만, 살짝 우려되는 부분은 Verification에 대해 Exception 클래스를 분리해두셨는데 정작 IllegalStateException이나 JwtException으로 throwing 하는 부분만 있는 것 같습니다!

해당 부분에서 메서드 throws 부분이나, throw문으로 던지는 부분을 미리 정의해둔 Verification Exception으로 감싸거나, 변경하면 좀 더 예외처리에 강점있는 코드가 될 것 같습니다!

polyglot-k reviewed (approved):
일단 승인해드릴게요 추후에 다시 확인할게요
